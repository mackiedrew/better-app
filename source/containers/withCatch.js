import React, { Component } from "react"

import { crashlytics } from "../firebase"

export default (ErrorComponent, errorMessage) => Child =>
  class extends Component {
    constructor(props) {
      super(props)
      this.state = {
        hasError: false,
        errorObject: null,
        errorInfo: null,
      }
    }
    componentDidCatch = (error, info) => {
      crashlytics.log(error)
      crashlytics.log(info)
      crashlytics.recordError(1, "withCatch")
      /* eslint-disable-next-line react/no-set-state */
      this.setState(() => ({
        hasError: true,
        errorObject: null,
        errorInfo: null,
      }))
    }
    render = () =>
      this.state.hasError ? (
        <ErrorComponent errorMessage={errorMessage} {...this.state} {...this.props} />
      ) : (
        <Child {...this.props} />
      )
  }
