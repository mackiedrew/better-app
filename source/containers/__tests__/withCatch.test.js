jest.mock("../../firebase", () => ({
  crashlytics: {
    log: jest.fn(),
    recordError: jest.fn(),
  },
}))

import React from "react"
import { Text } from "react-native"

import Renderer from "react-test-renderer"

import withCatch from "../withCatch"

const ErrorComponent = () => <Text>Error</Text>
const TestComponent = () => <Text>Test</Text>
const BrokenComponent = () => {
  /* eslint-disable-next-line fp/no-throw */
  throw new Error("Some Error!")
}

describe("withCatch", () => {
  it("matches snapshots with ErrorComponent", () => {
    const ResultComponent = withCatch(ErrorComponent)(TestComponent)
    const Render = Renderer.create(<ResultComponent />).toJSON()
    expect(Render).toMatchSnapshot()
  })
  it("matches snapshots with ErrorComponent and a throw", () => {
    const ResultComponent = withCatch(ErrorComponent, "Custom Error")(TestComponent)
    const Render = Renderer.create(<ResultComponent />).toJSON()
    expect(Render).toMatchSnapshot()
  })
  it("matches snapshots with ErrorComponent and errorMessage", () => {
    const ResultComponent = withCatch(ErrorComponent)(BrokenComponent)
    const Render = Renderer.create(<ResultComponent />).toJSON()
    expect(Render).toMatchSnapshot()
  })
  it("matches snapshots with ErrorComponent and errorMessage and a throw", () => {
    const ResultComponent = withCatch(ErrorComponent, "Custom Error")(BrokenComponent)
    const Render = Renderer.create(<ResultComponent />).toJSON()
    expect(Render).toMatchSnapshot()
  })
})
