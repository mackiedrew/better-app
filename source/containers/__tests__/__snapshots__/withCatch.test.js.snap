// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withCatch matches snapshots with ErrorComponent 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
  Test
</Text>
`;

exports[`withCatch matches snapshots with ErrorComponent and a throw 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
  Test
</Text>
`;

exports[`withCatch matches snapshots with ErrorComponent and errorMessage 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
  Error
</Text>
`;

exports[`withCatch matches snapshots with ErrorComponent and errorMessage and a throw 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
>
  Error
</Text>
`;
